src/pipeline.rs

```diff
@@ -207,25 +247,64 @@ async fn handle_insight(
     skip_to: Option<usize>,
     insight: Event,
     pipeline: &mut ExecutableGraph,
-    onramps: &Onramps,
+    inputs: &Inputs,
 ) {
     let insight = pipeline.contraflow(skip_to, insight);
     if insight.cb != CbAction::None {
-        for (_k, (send, o)) in onramps {
+        let mut input_iter = inputs.values();
+        let first = input_iter.next();
+        for (send, input) in input_iter {
+            let is_cb = insight.cb.is_cb();
+            if *send || is_cb {
+                if let Err(e) = match input {
+                    Input::Onramp(addr) => addr
+                        .send(onramp::Msg::Cb(insight.cb, insight.id.clone()))
+                        .await
+                        .map_err(Error::from),
+                    Input::Pipeline(addr) => addr.send_insight(insight.clone()).await,
+                    // linked offramps dont support insights yet, although they definitely should
+                    Input::LinkedOfframp(_addr) => {
+                        warn!(
+                            "Linked offramps don't support insights/GD yet, we are deeply sorry!"
+                        );
+                        Ok(())
+                    }
+                } {
+                    error!(
+                        "[Pipeline] failed to send insight to input: {} {:?}",
+                        e, &input
+                    );
+                }
+            }
+        }
+        if let Some((send, input)) = first {
             if *send || insight.cb.is_cb() {
-                if let Err(e) = o
-                    .send(onramp::Msg::Cb(insight.cb, insight.id.clone()))
-                    .await
-                {
-                    error!("[Pipeline] failed to send to onramp: {} {:?}", e, &o);
+                if let Err(e) = match input {
+                    Input::Onramp(addr) => addr
+                        .send(onramp::Msg::Cb(insight.cb, insight.id))
+                        .await
+                        .map_err(Error::from),
+                    Input::Pipeline(addr) => addr.send_insight(insight).await,
+                    // linked offramps dont support insights yet, although they definitely should
```

Comment:
Make this a `TODO`

Reply:
Yes, definitely! Will ticket this as well